networkx/algorithms/wiener.py

```diff
@@ -224,3 +224,61 @@ def gutman_index(G, weight=None):
     spl = nx.shortest_path_length(G, weight=weight)
     d = dict(G.degree, weight=weight)
     return sum(dist * d[u] * d[v] for u, vinfo in spl for v, dist in vinfo.items()) / 2
+
+
+@nx.utils.not_implemented_for("directed")
+@nx.utils.not_implemented_for("multigraph")
+@nx._dispatchable(edge_attrs="weight")
+def hyper_wiener_index(G, weight=None):
+    r"""Returns the Hyper-Wiener index of the graph G.
+
+    The Hyper-Wiener index of a connected graph G is defined as
+
+    .. math::
+        WW(G) = \frac{1}{2} \sum_{u,v \in V(G)} (d(u,v) + d(u,v)^2)
+
+    where `d(u,v)` is the shortest-path distance between nodes `u` and `v`.
+
+    Parameters
+    ----------
+    G : NetworkX graph
+        An undirected, connected graph.
+
+    weight : string or None, optional (default: None)
+        The edge attribute to use for calculating shortest-path distances.
+        If None, all edges are considered to have a weight of 1.
+
+    Returns
+    -------
+    float
+        The Hyper-Wiener index of the graph G.
+        Returns float("inf") if the graph is not connected.
+
+    References
+    ----------
+    .. [1] M. Randić, "Novel molecular descriptor for structure-property studies,"
+           Chemical Physics Letters, vol. 211, pp. 478-483, 1993.
+    .. [2] `Wikipedia: Hyper-Wiener Index <https://en.wikipedia.org/wiki/Hyper-Wiener_index>`_
+
+    Examples
+    --------
+    >>> G = nx.path_graph(4)
+    >>> nx.hyper_wiener_index(G)
+    30.0
+
+    >>> G = nx.cycle_graph(4)
+    >>> nx.hyper_wiener_index(G)
+    20.0
+    """
+    if len(G) < 2:
+        return 0.0
```

Comment:
This condition should be removed. There are instances where this would not be true, e.g. weighted self-loops.